reference/src/layout/arrays.md

```diff
@@ -0,0 +1,61 @@
+# Layout of Rust array types
+
+Array types, `[T; N]`, store `N` values of type `T` with a constant
+_stride_, where _stride_ is the distance between each pair of consecutive values
+within the array.
+
+The _offset_ of the first array element is `0`, that is, a pointer to the array
+and a pointer to its first element point to the same memory address.
+
+The _stride_ of the array is computed as the _size_ of the element type rounded up
+to the next multiple of the _alignment_ of the element type.
+
+When the element _size_ is a multiple of the element's _alignment_, then `stride
+== size`, and the elements are laid out contiguously in memory, e.g., `[u8; 4]`.
+In this case, the size of the array can be computed as `size_of::<T>() * N`[^1].
+
+> **Note:** In the current Rust implementation, _size_ is always a multiple of
+> the element's _alignment_, and therefore `stride == size` always holds. This
+> is, however, not guaranteed by the [layout of structs and tuples].
+
+[^1]: The alignment of the array is, however, unspecified. For example, the
+[SysV AMD64 ABI] requires array arguments to be at least 16 byte aligned to
+allow the use of SSE instructions.
```

Comment:
I don't think this is quite right for this context. I believe you are referring to

> An array uses the same alignment as its elements, except that a local or globalarray variable of length at least 16 bytes or a C99 variable-length array variablealways has alignment of at least 16 bytes.

That is quite different from the ABI alignment that `std::mem::align_of[_val]` and all these layout documents describe, which applies regardless of storage location and is also a restriction on unsafe code that allocates memory/constructs objects. But since globals and locals are entirely allocated by the compiler, overaligning them need not impact the alignment.

Reply:
I see what you mean. So what do we know about the array alignment ? Is it the same as the alignment of its element? Or can it be larger ?

For example, supposing I wanted to interface with C via an array of `repr(C)` types

```rust
extern {
    fn foo(x: *mut [ReprC; 8]) -> *mut [ReprC; 8];
}
```

is it enough if C returns me an array where the alignment of the array matches the alignment of `ReprC`, or can Rust decide that this array has a larger alignment requirement making this illegal ?

Reply:
If Rust arrays of ReprC types are layout compatible with C arrays, then the alignment requirements of these arrays would need to match those of C. But I'm not certain that this is (or should be) the case of arrays of `repr(Rust)` types.

Reply:
Rust unilaterally imposing larger ABI alignment requirements on arrays than the corresponding C array type would break FFI as you say, so clearly we can't do that. For Rust types that aren't suitable for FFI (e.g., unspecified layout), there's no such external constraint, but I wasn't talking about those anyway. I'm just saying that the SysV aspect you refer to here has no relation as all to the `align_of` of (C-compatible) arrays. As such, it's potentially interesting trivia but not relevant to the layout document.

Reply:
So I've removed that part, and I've written down the guarantees about array alignment down.